### PR TITLE
Add support for custom headers in the agent proxy

### PR DESCRIPTION
**Pull request by costinm**
Change-Id: I0dee2da459f05ee5a32a23438a03ab516877cd6f

Envoy - when using google-grpc - supports adding custom headers. 
This is needed when the XDS server is running behind external load balancers, or needs special handling.

Note that "ClusterID" is already added to headers. 

Using the same mechanism we use for node metadata - env variables - this is not a candidate for 
ProxyConfig because we can't guarantee support for all cases. 

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — APPROVED** [review]

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -57,6 +57,9 @@ const (
 	trustworthyJWTPath = "./var/run/secrets/tokens/istio-token"
 	localHostIPv4      = "127.0.0.1"
 	localHostIPv6      = "[::1]"
+
+	// Similar with ISTIO_META_, which is used to customize the node metadata - this customizes extra header.
```
Comment (yangminzhu) [inline pilot/cmd/pilot-agent/main.go:61]: can you add some more comments about the format of the header value and how it is being used.

Comment (costinm) [reply pilot/cmd/pilot-agent/main.go:60]: Nothing special - in envoy you can add some headers, we lost this ability here. Each server may have its own extra headers ( apikeys, custom params, etc ). The format is similar with ISTIO_META_ - prefix is removed, the rest is header name and value.

Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/main.go:60]: I think it just provides an alternative way to pass metadata. Besides for auth, is there any use case?

Comment (howardjohn) [reply pilot/cmd/pilot-agent/main.go:60]: @hzxuzhonghu we have also looked into a gateway that does  header based routing to different Istiod instances. but primary case is auth

**Pull request merged**